src/rpp/mutex.cpp

```diff
@@ -75,5 +83,160 @@ namespace rpp
         if (CRITICAL_SECTION* cs = GET_RMUTEX())
             LeaveCriticalSection(cs);
     }
+#elif RPP_FREERTOS
+    /////////////////////////////////////////////////////////////////
+
+    mutex::mutex()
+    {
+        mtx.ctx = xSemaphoreCreateBinary(); // binary semaphore is faster than mutex semaphore
+        xSemaphoreGive((SemaphoreHandle_t)mtx.ctx); // Initialize to available
+    }
+
+    mutex::~mutex()
+    {
+        if (mtx.ctx)
+        {
+            vSemaphoreDelete((SemaphoreHandle_t)mtx.ctx);
+            mtx.ctx = nullptr;
+        }
+    }
+
+    bool mutex::try_lock()
+    { 
+        if (xPortIsInsideInterrupt())
+        {
+            BaseType_t higherPriorityTaskWoken = pdFALSE;
+            bool r = xSemaphoreTakeFromISR((SemaphoreHandle_t)mtx.ctx, &higherPriorityTaskWoken) == pdTRUE;
+            portYIELD_FROM_ISR(higherPriorityTaskWoken);
+            return r;
+        }
+
+        return xSemaphoreTake((SemaphoreHandle_t)mtx.ctx, 0) == pdTRUE;
+    }
+    
+    void mutex::lock()
+    {
+        if (xPortIsInsideInterrupt())
+        {
+            BaseType_t higherPriorityTaskWoken = pdFALSE;
+            xSemaphoreTakeFromISR((SemaphoreHandle_t)mtx.ctx, &higherPriorityTaskWoken);
+            portYIELD_FROM_ISR(higherPriorityTaskWoken);
+        }
+
+        else
+            xSemaphoreTake((SemaphoreHandle_t)mtx.ctx, portMAX_DELAY);
+    }
```

Comment:
Misleading/confusing whitespace here. Don't separate `else` clauses like this, since it can lead to confusion. If we look at JSF and MISRA then they use the "Structured" (single flow return) coding style, which has its benefits for readability and maintainability:
```cpp
        if (xPortIsInsideInterrupt())
        {
            BaseType_t higherPriorityTaskWoken = pdFALSE;
            xSemaphoreTakeFromISR((SemaphoreHandle_t)mtx.ctx, &higherPriorityTaskWoken);
            portYIELD_FROM_ISR(higherPriorityTaskWoken);
        }
        else
        {
            xSemaphoreTake((SemaphoreHandle_t)mtx.ctx, portMAX_DELAY);
        }
```